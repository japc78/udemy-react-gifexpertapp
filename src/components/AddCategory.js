import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ( { setCategories } ) => {

	const [inputValue, setInputValue] = useState('');

	const handleInputChange = (e) => {
		setInputValue(e.target.value);
		// console.log(e.target);
		// Comprobamos a la hora de hacer test para por el manejador
		// console.log('handleInputChange llamado');
	}

	const handleSubmit = (e) => {
		e.preventDefault();
		// Comprobamos a la hora de hacer test que simula el submit
		// console.log('Pass: handleSubmit', inputValue);

		if (inputValue.trim().length > 2) {
			setCategories(c => [inputValue, ...c]);
			setInputValue('');
			// console.log('Submit hecho!');
		}

	}

	return (
		<form onSubmit= {handleSubmit}>
			{/* Se a√±ade para comprobar que el valor del input cambio en el evento onChange */}
			<p>{ inputValue }</p>
			<h2>Add Category</h2>
			<input
				type="text"
				value = {inputValue}
				onChange = {handleInputChange}
			/>
		</form>
	)
}

AddCategory.propTypes = {
	setCategories : PropTypes.func.isRequired
}
