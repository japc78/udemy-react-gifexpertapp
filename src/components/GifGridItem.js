import React from 'react'

// Con props ya pasadas de manera independiente.
// export const GifGridItem = ( props ) => {
// 	console.log(props);

// De una manera mas amigable para la lectura del codigo
export const GifGridItem = ( {id, title, url} ) => {
	console.log({id, title, url});

	return (
		<div>
			<img src={url} alt={title}/>
			<p>{title}</p>
		</div>
	)
}
