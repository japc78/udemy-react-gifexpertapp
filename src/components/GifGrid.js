import React, { useState, useEffect } from 'react'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ( {category }) => {

	const [images, setImages] = useState([]);

	// Es igual ComponentDidmount del modo antiguo de React con Clases, es decir el componente solo se va renderizar la primera vez.
	// Se utiliza el hook useEffect que como primer parametro es un callback que sera la funcion
	// que hace la peticion y como segundo parametro un array de dependencias, en este caso vacio
	// para que solo se dispare una sola vez.
	useEffect(() => {
		getGifs();
	},[]);

	const getGifs = async () => {
		const url = 'https://api.giphy.com/v1/gifs/search?q=Regular show&limit=10&api_key=LtkGU0ZbRcWCOsCJNT1RceZ9iE6w2Pw6'
		const respuesta = await fetch( url );
		const { data } = await respuesta.json();

		const gifs = data.map (img => {
			return {
				id: img.id,
				title: img.title,
				//la ? indica que si existe la url la traiga sino que no la pinte.
				url: img.images?.downsized_medium.url
			}
		})

		// console.log(data);
		// console.log(gifs);
		setImages( gifs );
	}

	// getGifs();

	return (
		<div>
			<h3>{ category }</h3>
				{
					// images.map(img => (
					// 	<li key = {img.id}> {img.title} </li>
					// ))

					// Con destructuracion
					images.map( img => (
						<GifGridItem
							key = {img.id}
							// img = {img}

							// Se envian las props diretamente con desestructuracion de manera independiente
							{...img}
						/>
					))}
		</div>
	)
}
