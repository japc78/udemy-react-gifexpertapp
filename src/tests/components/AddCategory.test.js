import React from 'react';
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import { AddCategory } from '../../components/AddCategory';

describe('Pruebas en componente <AddCategory />', () => {

	const setCategories = jest.fn();
	let wrapper = shallow( <AddCategory setCategories = {setCategories} />);
	let input = wrapper.find('input');

	//Resetar valores en cada test
	beforeEach( () => {
		jest.clearAllMocks();
		wrapper = shallow( <AddCategory setCategories = {setCategories} />);
		input = wrapper.find('input');
	});


	test('Debe de mostrarse correctamente', () => {
		expect( wrapper ).toMatchSnapshot();
	})

	test('Debe de cambiar la caja de texto', () => {
		const value = 'Hola Mundo';

		//Le pasamos el valor value al evento llamado en el componente AddCategory en el evento onChange, se simula que se le envia un valor.
		//Es lo mismo que
		// target: {
		// 	value: value
		// }
		input.simulate('change', { target : { value }});

		expect(wrapper.find('p').text().trim()).toBe( value );
	})

	test('No debe de postear la informacion con submit', () => {
		//Simulate primer parametro el evento, el segundo los parametros que van hacer transformados en el evento.
		//preventDefault(){} funcion enviada  dentro del objeto.

		wrapper.find('form').simulate('submit', {
			preventDefault: () => {}
		});


		// Lo mismo pero en forma corta
		// wrapper.find('form').simulate('submit', {preventDefault(){}});

		// Prueba en la que se indica que no llego a pasar el evento esperado, es decir llamar a la funcion setCategories.
		expect( setCategories ).not.toHaveBeenCalled();
	})

	test('Debe de llamar el setCategories y limpiar la caja de texto', () => {
		// 1. Simular el inputChange
		const value = 'Rick and Morty';
		input.simulate('change', { target : { value }});

		// 2. Simular el submit
		wrapper.find('form').simulate('submit', { preventDefault(){} });

		// 3. setCategories se debe haber llamado
		// expect( setCategories ).toHaveBeenCalled();

		// Se comprueba que es llamada solo una vez
		expect( setCategories ).toHaveBeenCalledTimes(1);

		// Se comprueba que es llamada como una funcion
		expect( setCategories ).toHaveBeenCalledWith( expect.any(Function));


		// 4. el valor del input debe estar vacio
		// console.log(input.text());
		expect(input.prop('value')).toBe('');

	})



})
