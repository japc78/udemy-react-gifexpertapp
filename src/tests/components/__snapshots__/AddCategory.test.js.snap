// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <AddCategory /> Debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <h2>
    Add Category
  </h2>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
